test/app/post_stream_test.dart

```diff
@@ -0,0 +1,165 @@
+        getPostsProvider,
+        listener.call,
+        fireImmediately: true,
+      );
+
+      verifyNever(() => listener(null, const AsyncData<List<Post>>([])));
+
+      // Act
+      final posts = await container.read(getPostsProvider.future);
+
+      // Assert
+      verifyInOrder([
+        () => listener(null, const AsyncLoading<List<Post>>()),
+        () => listener(
+              const AsyncLoading<List<Post>>(),
+              any(that: isA<AsyncData<List<Post>>>()),
+            ),
+      ]);
+
+      expect(posts.length, 2);
+    });
+    test('Test posts failure', () async {
+      final postClient = MockPostClient();
+      final exception = Exception('Posts connection failed');
+      when(() => postClient.getPosts(100)).thenThrow(exception);
+
+      final container = createContainer(postClient);
+      final listener = Listener<AsyncValue<List<Post>>>();
+
+      container.listen<AsyncValue<List<Post>>>(
+        getPostsProvider,
+        listener.call,
+        fireImmediately: true,
+      );
+
+      const data = AsyncData<List<Post>>([]);
+      verifyNever(() => listener(null, data));
+
+      await expectLater(
+        () async => container.read(getPostsProvider.future),
+        throwsA(isA<Exception>()),
+      );
+      // verify
+      verifyInOrder([
+        () => listener(null, const AsyncLoading<List<Post>>()),
+        () => listener(
+              // ignore: strict_raw_type
+              any(that: isA<AsyncLoading>()), any(that: isA<AsyncError>()),
+            ),
+      ]);
+      verifyNoMoreInteractions(listener);
+    });
+  });
+  group('Test Post Stream Widget', () {
+    testWidgets('success', (tester) async {
+      final postClient = MockPostClient();
+
+      final data = <Post>[
+        const Post(
+          id: 0,
+          title: Renderable(rendered: 'Create todo list app'),
+          excerpt: Renderable(rendered: 'Create todo list app'),
+        ),
+        const Post(
+          id: 1,
+          title: Renderable(rendered: 'Create todo list app'),
+          excerpt: Renderable(rendered: 'Create todo list app'),
+        ),
+      ];
+
+      when(() => postClient.getPosts(100))
+          .thenAnswer((invocation) async => Future.value(data));
+
+      await pumpApp(tester, postClient);
+
+      expect(find.byType(CircularProgressIndicator), findsOneWidget);
+
+      await tester.pumpAndSettle(const Duration(seconds: 1));
+      expect(find.byType(PostCell), findsExactly(2));
+    });
+    testWidgets('failure', (tester) async {
+      final postClient = MockPostClient();
+      final exception = Exception('Posts connection failed');
+      when(() => postClient.getPosts(100)).thenThrow(exception);
+
+      await pumpApp(tester, postClient);
+
+      expect(find.byType(CircularProgressIndicator), findsOneWidget);
+
+      await tester.pumpAndSettle(const Duration(seconds: 2));
+      expect(find.text('Exception: Posts connection failed'), findsOneWidget);
```

Comment:
@alexdivadi  plz make a jira ticket fo rus to do robust error handling

Reply:
Linked here: [WA-28](https://anirac-tech.atlassian.net/browse/WA-28?atlOrigin=eyJpIjoiNWZkZDdmNWRiMDc1NGUxZjk2YzAwOTEyZDExZmVkMzYiLCJwIjoiaiJ9)